ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp

```diff
@@ -203,6 +204,14 @@ static INT CALLBACK GetDirectoryAltProc(HWND hwnd, UINT uMsg, LPARAM lp, LPARAM
   return 0;
 }
 
+typedef basic_string<WCHAR> tstring;
+tstring widen(const string &str) {
+  // Number of shorts will be <= number of bytes; add one for null terminator
+  const size_t wchar_count = str.size() + 1;
+  vector<WCHAR> buf(wchar_count);
+  return tstring{buf.data(), MultiByteToWideChar(CP_UTF8, 0, str.c_str(), -1, buf.data(), wchar_count)};
```

Comment:
The narrowing is in `wchar_count`. If we want to be all proper, we can do something like this:

```C++
#  ifdef DEBUG_MODE
  if (wchar_count > std::numeric_limits<int>::max) show_error("Message characters will be discarded.", false);
#  endif
```

Then just cast wchar_count to `(int)`.

In my opinion, that warning is kind of silly; it's hardly worth the code to generate it, because this only becomes relevant if the message is larger than two gigabytes. So I don't care if you want to include it or not. Go ahead and perform the cast either way, though.

If we also need to cast the `int`→`size_t` conversion, you can go ahead and store the return value in an int, verify it's greater than zero, then use it, else show a similar error. Again, I'm not opposed to this, but it should be pointless, as the return value is always zero or else the number of characters written.

Reply:
Ok I tried casting it to int, that doesn't fix the warning:
```cpp
Widget_Systems/Win32/dialogs.cpp: In function 'tstring widen(const string&)':
Widget_Systems/Win32/dialogs.cpp:212:102: warning: narrowing conversion of 'MultiByteToWideChar(65001u, 0ul, (& str)->std::basic_string<_CharT, _Traits, _Alloc>::c_str<char, std::char_traits<char>, std::allocator<char> >(), -1, buf.std::vector<_Tp, _Alloc>::data<wchar_t, std::allocator<wchar_t> >(), ((int)wchar_count))' from 'int' to 'std::basic_string<wchar_t>::size_type {aka unsigned int}' inside { } [-Wnarrowing]
   return tstring{buf.data(), MultiByteToWideChar(CP_UTF8, 0, str.c_str(), -1, buf.data(), wchar_count)};
                                                                                                      ^
```

Reply:
Ok, I got it, I added both just in case:
82a78205ef9f0d0142c65729aa5035a06d00d0ba